tests/contracts_expects_3.nit

```diff
@@ -0,0 +1,47 @@
+# This file is part of NIT ( http://www.nitlanguage.org ).
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+class MyClass
+	fun foo(x: Int)
+	do
+		print "Good"
+	end
+end
+
+class SubClass
+	super MyClass
+
+	redef fun foo(x: Int)
+	is
+		expects(x == 0)
+	do
+		if x != 0 then print "FAIL"
+	end
+end
+
+class SubSubClass
+	super SubClass
+
+	redef fun foo(x: Int)
+	do
+		if x != 0 then print "FAIL"
+	end
+end
+
+var first = new MyClass
+first.foo(1) # OK
+var sub = new SubClass
+sub.foo(0) # OK
+var subsub = new SubSubClass
+subsub.foo(1) # FAIL
```

Comment:
Can we have a few more complete/complex tests?

Like:
* using both annotations in the same method
* using them with an abstract method
* using them in an abstract class / interface
* accessing attributes
* annotate constructors and factories
* annotate auto getters/setters

Reply:
Also, how do you access the returned value from the post condition?

I think we also need a way to refer to the value of something during the pre condition:

```nit
class Foo
    var i = 0

    fun inc
    is ensures(i == pre(i) + 1)
    do
        i += 1
    end
```

Reply:
I just added new tests. For tests with the class / abstract method, I had not thought about this use case. For information, currently the abstract methods and methods in an interface provide the implementation of the verification method.

get/set: For the moment this is not possible. It is possible to annotate the attributes with `expects` and` ensures` but it will not do anything.

Refer value: For the possibility of adding a `cache` to keep the value passed as input it will add this in a future version.